File: 2615-sum-of-distances/2615-sum-of-distances.go
func distance(nums []int) []int64 {
	arr := make([]int64, len(nums))
	m := map[int][]int{}
	for i, num := range nums {
		m[num] = append(m[num], i)
	}

	for _, a := range m {
		if len(a) > 1 {
			sumPos := 0
            count := 0
			for i := 1; i < len(a); i++ {
				sumPos += a[i]
                count++
			}

            sumNeg := 0
			for i := range a {
				arr[a[i]] = int64((a[i] * (i - count)) - sumNeg + sumPos)
                if i < len(a) - 1 {
                    sumNeg += a[i]
                    sumPos -= a[i+1]
                    count--
                }
			}
		}
	}

    return arr
}

// array, hash map, math, prefix sum
// time: O(n)
// space: O(n)

// tips
// go grab some paper and a pen :)

File: 2615-sum-of-distances/2615-sum-of-distances.py
class Solution:
    def distance(self, nums: List[int]) -> List[int]:
        ans = [0] * len(nums)

        m = {}
        for i, num in enumerate(nums):
            if num not in m:
                m[num] = []
            m[num].append(i)

        for k in m:
            if len(m[k]) > 1:
                arr = m[k]
                
                sum_pos, sum_neg = 0, 0
                count = 0
                for i in range(1, len(arr)):
                    sum_pos += arr[i]
                    count += 1
                
                for i in range(len(arr)):
                    ans[arr[i]] = (arr[i] * (i - count)) - sum_neg + sum_pos
                    if i < len(arr) - 1:
                        count -= 1
                        sum_neg += arr[i]
                        sum_pos -= arr[i + 1]
        
        return ans

# array, hash map, math, prefix sum
# time: O(n)
# space: O(n)

# tips
# go grab some paper and a pen :)